Apply precision in compress bank last

cdcapano opened the pull request:
Currently, if `pycbc_compress_bank` is run with `--precision single`, the entire compression algorithm is run using single precision. This limits what tolerance can be set; we've found that specifying a precision of 0.0001 can cause problems for longer waveforms, since the mismatch cannot actually be measured to that high of a precision if the waveforms are generated using single precision. This patch fixes that by running the compression algorithm at double precision, and only saving the results to single precision at the end.

I have noticed that when doing double precision, fewer sample points are returned by the spa compression algorithm. For example, in the template that was proving to be problematic with a tolerance of 0.0001, ~1700 points are selected when using double precision, whereas ~1900 points were selected when using single precision. This is probably due to accumulated errors in the spa algorithm. Since this patch causes the spa algorithm to always be run at double precision, this means the smaller number of points will be used even when saving as single precision. I think this should be ok, but this should be tested to ensure the mismatch is still below the desired tolerance when using single precision.

Putting on hold for now, as this relies on #1491.

soumide1102 commented on pycbc/waveform/compress.py:
```diff
@@ -726,17 +749,28 @@ def write_to_hdf(self, fp, template_hash, root=None):
+        elif precision == 'double' and self.precision == 'single':
+            raise ValueError("cannot cast single precision to double")
+        outdtype = _real_dtypes[precision]
         group = '%scompressed_waveforms/%s' %(root, str(template_hash))
```
@cdcapano Line 630 sets `self.precision=precision` . Thus, when CompressedWaveform is generated by `compressed_waveform` it would save `self.precision` to be the same as the user input `args.precision` from `pycbc_compress_bank` . Again, here in `write_to_hdf`, `precision` would be the same as `args.precision` from `pycbc_compress_bank`. So, would it help to compare `precision` and `self.precision` , because aren't they going to be the same always? So instead, how about replacing lines 762-765 with just :
```
if precision is None:
    precision = 'single'
```
The compressed waveform points would be generated as `double` by default. So if the user provides `--precision` as `single` in `pycbc_compress_bank`, that gets passed as an argument to `write_to_hdf` and the compressed points get converted to `single` from `double` in line 766 above, when deciding `outdtype`. If the user provides `--precision` as `double`, points get saved as `double` and if `--precision` is `None`, again convert points generated as `double` to `single`.

soumide1102 commented on pycbc/waveform/compress.py:
```diff
@@ -749,18 +749,28 @@ def write_to_hdf(self, fp, template_hash, root=None):
+            Cast the saved parameters to the given precision before saving. If
+            None provided, will use whatever their current precision is. This
+            will raise an error if the parameters have single precision but the
+            requested precision is double.
```
Same comment as below. Aren't the points getting generated as `double` by default, as we are doing everything in `double` fixing `dtype=complex128` always earlier in the code? So would the sentence ```This will raise an error if the parameters have single precision but the requested precision is double.``` be relevant?

soumide1102 replied:
Also, should we change `precision=args.precision` to `precision='double'` when doing `psd=pycbc.psd.from_cli(...)` because the original waveform and the decompressed waveform would have `dtype = numpy.complex128`? So, to be safe, wouldn't it be better to make all the three arrays ie. the two waveforms and the psd have the same precision?

soumide1102 commented:
@cdcapano Have you got a chance to look at the changes I suggested in the comments?

cdcapano commented:
Just to note, the checking of the precision was intentional, since the precision specified when using `write_to_hdf` may not be the same as what was used when the class was initialized. I've reverted @soumide1102's commit. There was a bug in that the class defaults to initializing precision to None; if someone then used `write_to_hdf`, they'd get an error. I've fixed that by making the default precision `double` when the class is initialized.

soumide1102 commented:
@cdcapano This PR has conflicts. I can fix them and merge tomorrow, after I get back to Syracuse unless you have already fixed them by then.

soumide1102 commented:
@cdcapano `precision` should be set to `'double'` when getting `psd` using `pycbc.psd.from_cli` in compress.bank . Currently we have have it as `precision=args.precision` . Since we are doing all the calculations using double precision while compressing the points, this throws an error of mismatch of precision when calculating the overlap. I noticed this while compressing the O1 bank using the psd last week. So I changed I changed it to `precision=double` in my branch. Would you want to fix this in this PR or in a new PR? Other than that, I have fixed the conflicts in this branch and it is ready to be pushed.

cdcapano commented:
@soumide1102 Thanks for catching that. I've pushed the fix to set the psd to double. If you don't seen anything else wrong with it, please squash and merge once all of the checks have passed.

soumide1102 commented:
@cdcapano Thanks! Just another small thing that I noticed just now. Should it be `'double'` instead of `None` [here](https://github.com/ligo-cbc/pycbc/pull/1492/files#diff-ebbbac95a4e9dc2f846c380e46b630c3R579) ? Other than that everything looks good to me.

cdcapano commented:
@soumide1102 D'oh. Fixed.

soumide1102 commented:
Thanks @cdcapano . Merged.
